scripts/task/ExportDeliveryClassXmlResults.php

```diff
@@ -0,0 +1,167 @@
+    private $report;
+
+    protected function provideOptions()
+    {
+        return [
+            'deliveryClassUri' => [
+                'prefix' => 'u',
+                'longPrefix' => 'uri',
+                'required' => true,
+                'description' => 'Delivery Class URI.'
+            ],
+            'zipFilenamePrefix' => [
+                'prefix' => 'p',
+                'longPrefix' => 'filename_prefix',
+                'required' => false,
+                'defaultValue' => 'xml_results',
+                'description' => 'Prefix for ZIP file name'
+            ],
+        ];
+    }
+
+    protected function provideDescription()
+    {
+        return 'The script will generate a zip with xml results';
+    }
+
+    protected function provideUsage()
+    {
+        return [
+            'prefix' => 'h',
+            'longPrefix' => 'help',
+        ];
+    }
+
+    protected function run()
+    {
+        $this->report = Report::createInfo('Export started');
+
+        $class = $this->getOption('deliveryClassUri');
+        $zipFilenamePrefix = $this->getOption('zipFilenamePrefix');
+        $deliveryClass = $this->getClass($class);
+        $resultCount = 0;
+
+        $fileName = \tao_helpers_Display::textCleaner($zipFilenamePrefix) .'-'. time() . '.zip';
+        $path = \tao_helpers_File::concat([\tao_helpers_Export::getExportPath(), $fileName]);
+
+        if (!\tao_helpers_File::securityCheck($path, true)) {
+            throw new \Exception('Unauthorized file name');
+        }
+        if (file_exists($path)) {
+            unlink($path);
+        }
+        $zipArchive = new \ZipArchive();
+        if ($zipArchive->open($path, \ZipArchive::CREATE) !== true) {
+            throw new \Exception('Unable to create archive at ' . $path);
+        }
+
+        $deliveryNumber = 0;
+        $executionAcrossDeliveriesNumber = 0;
+        foreach ($deliveryClass->getInstances(true) as $delivery) {
+            $deliveryNumber++;
+            $this->report->add(Report::createInfo(sprintf('Starting to process delivery "%s"', $delivery->getLabel())));
+
+            try {
+                $executions = $this->getResultsService()->getResultsByDelivery($delivery);
+
+                $this->report->add(Report::createInfo(sprintf('%d results were found for delivery "%s"', count($executions), $delivery->getLabel())));
+
+                $executionNumber = 0;
+                foreach ($executions as $execution) {
+                    $executionNumber++;
+                    $executionAcrossDeliveriesNumber++;
+
+                    // flush the archive by creating the object ZipArchive from scratch every 100 iterations to
+                    // prevent ZipArchive from memory leaking (it builds up all the changes in memory that are not 'committed')
+                    if ($executionAcrossDeliveriesNumber % 100 === 0) {
+                        $zipArchive->close();
+                        $zipArchive = new \ZipArchive();
```

Comment:
Archive object creation takes place two times (here and on lines 78-81). May be we should create a separate method for this and called it createZipArchive(): \ZipArchive, for example?

Reply:
fixed